### ISSUE TITLE
ERROR: InvalidArgumentException: Unexpected data found

### ISSUE THREAD

**Issue by willdashwood** [openprovider/OP-WHMCS7#63]
Updated to the latest release and although it seems to have fixed the error where domain status wasn't being updated, I now get the following error when running DomainSync.php

`[WHMCS Application] ERROR: InvalidArgumentException: Unexpected data found. in /home/nginx/domains/domain.com/public/billing/vendor/nesbot/carbon/src/Carbon/Carbon.php:425 Stack trace: #0 /home/nginx/domains/domain.com/public/billing/vendor/whmcs/whmcs-foundation/lib/Model/AbstractModel.php(0): Carbon\Carbon::createFromFormat('Y-m-d H:i:s', '-0001-11-16 00:...') #1 /home/nginx/domains/domain.com/public/billing/vendor/illuminate/database/Eloquent/Model.php(2636): WHMCS\Model\AbstractModel->asDateTime('-0001-11-16 00:...') #2 /home/nginx/domains/domain.com/public/billing/vendor/illuminate/database/Eloquent/Model.php(2602): Illuminate\Database\Eloquent\Model->getAttributeValue('nextduedate') #3 /home/nginx/domains/domain.com/public/billing/vendor/whmcs/whmcs-foundation/lib/Model/AbstractModel.php(0): Illuminate\Database\Eloquent\Model->getAttribute('nextduedate') #4 /home/nginx/domains/domain.com/public/billing/vendor/illuminate/database/Eloquent/Model.php(3397): WH!
 MCS\Model\AbstractModel->getAttribute('nextduedate') #5 /home/nginx/domains/domain.com/public/billing/modules/registrars/openprovider/src/DomainSync.php(301): Illuminate\Database\Eloquent\Model->__get('nextduedate') #6 /home/nginx/domains/domain.com/public/billing/modules/registrars/openprovider/src/DomainSync.php(215): OpenProvider\WhmcsRegistrar\src\DomainSync->process_empty_next_due_dates() #7 /home/nginx/domains/domain.com/public/billing/modules/registrars/openprovider/cron/DomainSync.php(35): OpenProvider\WhmcsRegistrar\src\DomainSync->process_domains() #8 {main} {"exception":"[object] (InvalidArgumentException(code: 0): Unexpected data found. at /home/nginx/domains/domain.com/public/billing/vendor/nesbot/carbon/src/Carbon/Carbon.php:425)"} []`

**Comment by sdmpxs** [comment]
Just to cause some confusion, I'm not seeing this error. Are you seeing this in the terminal when running manually?

**Comment by willdashwood** [comment]
Both when running script directly over terminal and cron.

**Comment by willdashwood** [comment]
Still getting this error BTW. Fairly sure it's to do with unexpected data being returned by the OP API. I look at the bit of the error "createFromFormat('Y-m-d H:i:s', '-0001-11-16 00:...') " and wondering if it's all down to a date format error. Can this be looked into asap?

**Comment by soberoctober** [comment]
Hi @willdashwood 

Can you share a domain so we can replicate its state on our test environment?

**Comment by willdashwood** [comment]
I'm not sure what domain name is causing the issue. It's not mentioned in the debug info.

**Comment by willdashwood** [comment]
Had to roll back to the 3.1 release as this was preventing the domainsync.php script working and we use other registrars. 3.1 works fine in that it fixes the above issue. So clearly a bug has crept in between 3.1 and the latest release.

**Comment by willdashwood** [comment]
Having rolled back to 3.1 I've noticed the error cropping up again, so perhaps it is actually a fault with a domain in our account. Or rather, the API giving data for a domain in our account that the WHMCS module balk at. I'll raise a ticket and reference this thread.

**Comment by mitch2k** [comment]
Hi,

I'm having the same issue. suddenly I also have a number of domains with expiry date 0000-00-00. This causes WHMCS to invoice this domain at random dates...

Also, it seems that the error blocked the complete sync process causing all domains not to be updated anymore.

**Comment by soberoctober** [comment]
We have almost figured out this one.
It seems like certain API output data does affect the sync script.
This will be fixed in the nearest update.

**Comment by wmetge** [comment]
@mitch2k , @willdashwood would you be able to provide the full Openprovider module logs which result from a failed update attempt? If you send this information to the Openprovider support team, they'll be able to perform an investigation as to what may be causing the issue from the API side.

**Comment by willdashwood** [comment]
@wmetge Is there a specific module log for the OpenProvider module? I enabled System Module Debug Log in WHMCS and then ran domainsync.php but the only entries are for the Nominet module we also use. I still get the error:

`[WHMCS Application] ERROR: Exception: Route does not exist. in /home/nginx/domains/domain.com/public/billing/modules/registrars/openprovider/vendor/wedevelopcoffee/wpower/src/Core/Router.php:85 Stack trace: #0 /home/nginx/domains/domain.com/public/billing/modules/registrars/openprovider/vendor/wedevelopcoffee/wpower/src/Core/Launch.php(50): WeDevelopCoffee\wPower\Core\Router->findRoute('transferSync') #1 /home/nginx/domains/domain.com/public/billing/modules/registrars/openprovider/openprovider.php(258): WeDevelopCoffee\wPower\Core\Launch->output(Array, 'transferSync') #2 /home/nginx/domains/domain.com/whmcsdata/crons/domainsync.php(0): openprovider_TransferSync(Array) #3 {main} {"exception":"[object] (Exception(code: 0): Route does not exist. at /home/nginx/domains/domain.com/public/billing/modules/registrars/openprovider/vendor/wedevelopcoffee/wpower/src/Core/Router.php:85)"} []
`

**Comment by willdashwood** [comment]
When I run modules/registrars/openprovider/cron/DomainSync.php I get:

`[WHMCS Application] ERROR: InvalidArgumentException: Unexpected data found. in /home/nginx/domains/domain.com/public/billing/vendor/nesbot/carbon/src/Carbon/Carbon.php:425 Stack trace: #0 /home/nginx/domains/domain.com/public/billing/vendor/whmcs/whmcs-foundation/lib/Model/AbstractModel.php(0): Carbon\Carbon::createFromFormat('Y-m-d H:i:s', '-0001-11-16 00:...') #1 /home/nginx/domains/domain.com/public/billing/vendor/illuminate/database/Eloquent/Model.php(2636): WHMCS\Model\AbstractModel->asDateTime('-0001-11-16 00:...') #2 /home/nginx/domains/domain.com/public/billing/vendor/illuminate/database/Eloquent/Model.php(2602): Illuminate\Database\Eloquent\Model->getAttributeValue('nextduedate') #3 /home/nginx/domains/domain.com/public/billing/vendor/whmcs/whmcs-foundation/lib/Model/AbstractModel.php(0): Illuminate\Database\Eloquent\Model->getAttribute('nextduedate') #4 /home/nginx/domains/domain.com/public/billing/vendor/illuminate/database/Eloquent/Model.php(3397): WHMCS\Model\AbstractModel->getAttribute('nextduedate') #5 /home/nginx/domains/domain.com/public/billing/modules/registrars/openprovider/src/DomainSync.php(301): Illuminate\Database\Eloquent\Model->__get('nextduedate') #6 /home/nginx/domains/domain.com/public/billing/modules/registrars/openprovider/src/DomainSync.php(215): OpenProvider\WhmcsRegistrar\src\DomainSync->process_empty_next_due_dates() #7 /home/nginx/domains/domain.com/public/billing/modules/registrars/openprovider/cron/DomainSync.php(35): OpenProvider\WhmcsRegistrar\src\DomainSync->process_domains() #8 {main} {"exception":"[object] (InvalidArgumentException(code: 0): Unexpected data found. at /home/nginx/domains/domain.com/public/billing/vendor/nesbot/carbon/src/Carbon/Carbon.php:425)"} []
`

**Comment by wmetge** [comment]
@willdashwood It looks like it might be an issue with the content of the database. The error:
`/whmcs/whmcs-foundation/lib/Model/AbstractModel.php(0): Carbon\Carbon::createFromFormat('Y-m-d H:i:s', '-0001-11-16 00:...')`
appears to be trying to parse an invalid date.

Would you be able to extract a list of the due dates from the table  `tbldomains`. something like:
`select DISTINCT nextduedate from tbldomains;`

**Comment by willdashwood** [comment]
There's nothing that stands out to me anyway.

[nextduedates.txt](https://github.com/openprovider/OP-WHMCS7/files/4116318/nextduedates.txt)

**Comment by wmetge** [comment]
the first entry has date "0000-00-00". if you change this date, do you continue to have this error?

**Comment by willdashwood** [comment]
There are quite a few entries using "0000-00-00". What would you recommend updating them to?

**Comment by willdashwood** [comment]
`select id, nextduedate from tbldomains WHERE nextduedate = '0000-00-00';
/* Affected rows: 0  Found rows: 2,785  Warnings: 0  Duration for 1 query: 0.032 sec. (+ 0.062 sec. network) */`

**Comment by mitch2k** [comment]
From my end; I'm starting to see "0000-00-00" dates more and more. It started when I started using OP.
I think they should be updated to the correct date as from the API result..

**Comment by willdashwood** [comment]
Just to follow up, a lot of these domains with "0000-00-00" nextduedate aren't even associated with OpenProvider. We have other domain registrars active in our WHMCS installation. Even so, the plugin should be able to handle a date that's outside it's expected range. It shouldn't just crash out.

**Comment by wmetge** [comment]
We've added a fix which checks for a null date (0000-00-00) and replaces it with a zero date (1970-01-01)

@mitch2k , could you send the Openprovider support department examples of the domains assigned to Openprovider which have null value set for the nextduedate?

**Comment by willdashwood** [comment]
Thanks, I've just tried out the latest v3.2 branch and it runs without error. Furthermore, the email that was sent showed next due dates of "-0001-11-30" which is most bizzare.

![sg20200206](https://user-images.githubusercontent.com/4964200/73939858-3fe4f600-48e2-11ea-8f10-239ee3e9458e.png)

I did a bit of digging and the first of these 2 domains had become orphaned somehow. There was a record in tbldomains but the client it belonged to had been deleted. Almost certainly this would have been from when we imported data from ClientExec to WHMCS years ago.

This might be going off topic here but can I just clarify what cron jobs with this latest release? Do we still need to run "modules/registrars/openprovider/cron/DomainSync.php" or does it tie into the standard WHMCS (crons/cron.php) cron script these days? I ask because I'm noticing that the expiry and next due dates appear to be being set in the default "WHMCS Domain Synchronisation Cron Report".

![sg20200206_2](https://user-images.githubusercontent.com/4964200/73940654-bd5d3600-48e3-11ea-8dd0-62e01f5fcca6.png)

That said, it does seem that the domains in that email that are with OpenProvider always output the notice that the expiry and next due date were updated, even though they weren't (i.e. they were already set to that before). Is that a small, minor bug? Don't really want the notification email cluttered up if now changes were actually made.

Thanks
